### PR TITLE
Integration & Derivation Modules

### LINKED ISSUE #32: Integration and derivative module

### PR DESCRIPTION
**Pull request by mayankDhiman**
**Changelog:**
- Added differentiation & Integration Methods to _Function_ classes _(including trigonometry)_
- Added Product Rule (Differentiation) and By-Parts (Integrations) modules
- Unit tests for Product Rule & for Integration & Differentiation methods for function classes

**Issues:**
- Should fix #32

### CODE REVIEW

File: visma/io/parser.py [hunk 1]
```diff
@@ -221,6 +222,18 @@ def tokensToString(tokens):
                 if token.power != 1:
                     tokenString += "^" + "(" + str(token.power) + ")"
                 tokenString += "(" + tokensToString([token.operand]) + ")"
+        elif isinstance(token, Sine) or isinstance(token, Cosine) or isinstance(token, Tangent) or isinstance(token, Cosecant) or isinstance(token, Secant) or isinstance(token, Cotangent):
```
Comment (8hantanu) [inline visma/io/parser.py]: Create a superclass trigonometric and make sin, cos, etc as sub-classes. 
```suggestion
        elif isinstance(token, Trigonometric):
```

Comment (8hantanu) [inline visma/io/parser.py]: Create a superclass trigonometric and make sin, cos, etc as sub-classes. 
```suggestion
        elif isinstance(token, Trigonometric):
```

File: visma/simplify/simplify.py [hunk 2]
```diff
@@ -259,14 +260,24 @@ def expressionSimplification(tokens_now, scope, tokens1):
...
         comments += [['Opening up all the brackets']]
-    if scope == []:
-        simToks, availableOperations, token_string, animExtra, commentExtra = simplifification(simToks)
-        animExtra.pop(0)
-        animation += animExtra
-        comments += commentExtra
+
+    trigonometricError = False
+    for tk in simToks:
+        if isinstance(tk, Sine) or isinstance(tk, Cosine) or isinstance(tk, Tangent) or isinstance(tk, Cosecant) or isinstance(tk, Secant) or isinstance(tk, Cotangent):
```
Comment (8hantanu) [inline visma/simplify/simplify.py]:
```suggestion
        if isinstance(tk, Trigonometric):
```

File: visma/calculus/differentiation.py [hunk 3]
```diff
@@ -61,43 +61,47 @@ def differentiateTokens(funclist, wrtVar):
...
-                                if(funcCopy.power[i] == 0):
-                                    del funcCopy.power[i]
-                                    del funcCopy.value[i]
-                                    if funcCopy.value == []:
-                                        funcCopy.__class__ = Constant
-                                        funcCopy.value = funcCopy.coefficient
-                                        funcCopy.coefficient = 1
-                                        funcCopy.power = 1
-                        commentsNew[0] += r"$" + r"= " + funcCopy.__str__() + r"\ ;\ " + r"$"
+                    if isinstance(funcCopy, Sine) or isinstance(funcCopy, Cosine) or isinstance(funcCopy, Tangent) or isinstance(funcCopy, Cosecant) or isinstance(funcCopy, Secant) or isinstance(funcCopy, Cotangent) or isinstance(funcCopy, Logarithm) or isinstance(funcCopy, Variable) or isinstance(funcCopy, exponential):
```
Comment (8hantanu) [inline visma/calculus/differentiation.py]:
```suggestion
                    if isinstance(funcCopy, Trigonometric) or isinstance(funcCopy, Logarithm) or isinstance(funcCopy, Variable) or isinstance(funcCopy, Exponential):
```

File: visma/calculus/differentiation.py [hunk 4]
```diff
@@ -61,43 +61,47 @@ def differentiateTokens(funclist, wrtVar):
...
+                        funcCopy = funcCopy.differentiate(wrtVar)
```
Comment (8hantanu) [inline visma/calculus/differentiation.py:71]: Much better now.

File: visma/simplify/simplify.py [hunk 5]
```diff
@@ -259,14 +260,24 @@ def expressionSimplification(tokens_now, scope, tokens1):
     if expressionPresent:
         animation += [simToks]
         comments += [['Opening up all the brackets']]
-    if scope == []:
-        simToks, availableOperations, token_string, animExtra, commentExtra = simplifification(simToks)
-        animExtra.pop(0)
-        animation += animExtra
-        comments += commentExtra
+
+    trigonometricError = False
```
Comment (8hantanu) [inline visma/simplify/simplify.py:264]: What is this trigonometric error check for?

Comment (mayankDhiman) [reply visma/simplify/simplify.py:264]: The simplify module is not able to simplify the terms involving Trigonometric Expressions (like `sin(x)*x`). This has been marked as a TODO for now. When boolean `trigonometricError` turns true (i.e. some trigonometric term is there in the input), these tokens are left untouched by simplifying module.

File: visma/io/parser.py (hunk 1 above)
Comment (mayankDhiman) [reply visma/io/parser.py]: Done. Created an _empty_ Class _Trigonometric_ and all other Sine, Cosine etc. as its sub-classes

**Review by 8hantanu — APPROVED** [review]

**Pull request merged**